### ISSUE TITLE
maximum recursion depth exceeded while testing A770 with a simple MNIST Handwritten Digit Recognition Network

### ISSUE THREAD

**Issue by huhanwj** [intel/intel-extension-for-pytorch#277]
I tried a very simple network to see whether I can train some models on the Arc A770, and I got the error msg like this after training 1 epoch

It can run and do the training on BFloat16, but only just for 1 epoch.

SGD optimizer and basically the same code in the provided example are used.

```
RecursionError                            Traceback (most recent call last)
Cell In[11], line 2
      1 for epoch in range(1, n_epochs + 1):
----> 2     train(epoch)
      3     test()
      4 import matplotlib.pyplot as plt

Cell In[9], line 8, in train(epoch)
      6 network, optimizer = ipex.optimize(network, optimizer=optimizer, dtype=torch.bfloat16)
      7 for batch_idx, (data, target) in enumerate(train_loader):
----> 8     optimizer.zero_grad()
      9 ######################### code changes ######################### 
     10     data = data.to("xpu")

File ~/miniconda3/envs/pytorch-intel/lib/python3.9/site-packages/intel_extension_for_pytorch/optim/_optimizer_utils.py:41, in patch_zero_grad_for_master_weight_training.<locals>.zero_grad(self, set_to_none)
     39                 bf16_param.grad.requires_grad_(False)
     40             bf16_param.grad.zero_()
---> 41 self._original_zero_grad(set_to_none)

File ~/miniconda3/envs/pytorch-intel/lib/python3.9/site-packages/intel_extension_for_pytorch/optim/_optimizer_utils.py:41, in patch_zero_grad_for_master_weight_training.<locals>.zero_grad(self, set_to_none)
     39                 bf16_param.grad.requires_grad_(False)
     40             bf16_param.grad.zero_()
---> 41 self._original_zero_grad(set_to_none)

    [... skipping similar frames: patch_zero_grad_for_master_weight_training.<locals>.zero_grad at line 41 (9967 times)]

File ~/miniconda3/envs/pytorch-intel/lib/python3.9/site-packages/intel_extension_for_pytorch/optim/_optimizer_utils.py:41, in patch_zero_grad_for_master_weight_training.<locals>.zero_grad(self, set_to_none)
     39                 bf16_param.grad.requires_grad_(False)
     40             bf16_param.grad.zero_()
---> 41 self._original_zero_grad(set_to_none)

File ~/miniconda3/envs/pytorch-intel/lib/python3.9/site-packages/intel_extension_for_pytorch/optim/_optimizer_utils.py:30, in patch_zero_grad_for_master_weight_training.<locals>.zero_grad(self, set_to_none)
     28 def zero_grad(self, set_to_none: bool = False):
     29     for p in self.params_attr:
---> 30         if 'bf16_param' in self.params_attr[p]:
     31             bf16_param = self.params_attr[p]['bf16_param']
     32         if bf16_param.grad is not None:

File ~/miniconda3/envs/pytorch-intel/lib/python3.9/site-packages/torch/_tensor.py:651, in Tensor.__hash__(self)
    650 def __hash__(self):
--> 651     if has_torch_function_unary(self):
    652         return handle_torch_function(Tensor.__hash__, (self,), self)
    653     return id(self)

RecursionError: maximum recursion depth exceeded while calling a Python object
```

Environment:
*Installed via pip* following the document
* Ubuntu 20.04.5 LTS in WSL2
* torch.version == '1.10.0a0'
* ipex.version == '1.10.200+gpu'

`dpkg -l|grep intel-level-zero-gpu` gives

```
ii  intel-level-zero-gpu                  1.3.23726.1+i419~u20.04           amd64        Intel(R) Graphics Compute Runtime for oneAPI Level Zero.
```

**Comment by jingxu10** [comment]
Could you share the full training script that reproduces this issue?

**Comment by huhanwj** [comment]
> Could you share the full training script that reproduces this issue?

sure
``` python
import torch
import torchvision
from torch.utils.data import DataLoader
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import intel_extension_for_pytorch as ipex
import matplotlib.pyplot as plt

n_epochs = 100
batch_size_train = 64
batch_size_test = 1000
learning_rate = 0.01
momentum = 0.5
log_interval = 10
random_seed = 1
torch.manual_seed(random_seed)
train_loader = torch.utils.data.DataLoader(
    torchvision.datasets.MNIST('./data/', train=True, download=True,
                               transform=torchvision.transforms.Compose([
                                   torchvision.transforms.ToTensor(),
                                   torchvision.transforms.Normalize(
                                       (0.1307,), (0.3081,))
                               ])),
    batch_size=batch_size_train, shuffle=True)
test_loader = torch.utils.data.DataLoader(
    torchvision.datasets.MNIST('./data/', train=False, download=True,
                               transform=torchvision.transforms.Compose([
                                   torchvision.transforms.ToTensor(),
                                   torchvision.transforms.Normalize(
                                       (0.1307,), (0.3081,))
                               ])),
    batch_size=batch_size_test, shuffle=True)

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(784, 128)
        self.fc2 = nn.Linear(128, 64)
        self.fc3 = nn.Linear(64, 10)
        self.activate = nn.LeakyReLU()
        self.flatten = nn.Flatten()

    def forward(self, x):
        x = self.flatten(x)
        x = self.fc1(x)
        x = self.activate(x)
        x = self.fc2(x)
        x = self.activate(x)
        x = self.fc3(x)
        return x

network = Net()
optimizer = optim.SGD(network.parameters(), lr=learning_rate,
                      momentum=momentum)

train_losses = []
train_counter = []
test_losses = []
test_counter = [i*len(train_loader.dataset) for i in range(n_epochs + 1)]

def train(epoch):
    global network
    global optimizer
    network.train()
    network = network.to('xpu')
    network, optimizer = ipex.optimize(network, optimizer=optimizer, dtype=torch.bfloat16)
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        data = data.to("xpu")
        target = target.to("xpu")
        with torch.xpu.amp.autocast(enabled=True, dtype=torch.bfloat16):
            output = network(data)
            loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % log_interval == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))
            train_losses.append(loss.item())
            train_counter.append(
            (batch_idx*64) + ((epoch-1)*len(train_loader.dataset)))
            torch.save(network.state_dict(), './model.pth')
            torch.save(optimizer.state_dict(), './optimizer.pth')

def test():
    network.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data = data.to("xpu")
            target = target.to("xpu")
            output = network(data)
            test_loss += F.nll_loss(output, target, size_average=False).item()
            pred = output.data.max(1, keepdim=True)[1]
            correct += pred.eq(target.data.view_as(pred)).sum()
    test_loss /= len(test_loader.dataset)
    test_losses.append(test_loss)
    print('\nTest set: Avg. loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
    test_loss, correct, len(test_loader.dataset),
    100. * correct / len(test_loader.dataset)))

for epoch in range(1, n_epochs + 1):
    train(epoch)
    test()

fig = plt.figure()
plt.plot(train_counter, train_losses, color='blue')
plt.scatter(test_counter, test_losses, color='red')
plt.legend(['Train Loss', 'Test Loss'], loc='upper right')
plt.xlabel('number of training examples seen')
plt.ylabel('negative log likelihood loss')
plt.show()

```

I was first using jupyter notebook, but just right now I switch to normal python and I get this
```
[CRITICAL ERROR] Kernel '_ZTSZZN2at15AtenIpexTypeXPU4impl23launch_vec_kernel_sgdMWILi2EN3c108BFloat16EEEvRNS_6TensorES6_S6_dbS6_ddbdllENKUlRN2cl4sycl7handlerEE_clESA_EUlNS8_4itemILi1ELb1EEEE_' removed due to usage of FP64 instructions unsupported by the targeted hardware. Running this kernel may result in unexpected results.
```

**Comment by vishnumadhu365** [comment]
@huhanwj I could reproduce the original issue (related to max recursion) you had reported. Seems the issue was due to multiple invocation of ipex.optimize. Making the following change to the train function resolves the issue,
```python

def train(epoch):
    global network
    global optimizer
    network.train()
    if epoch == 1:
      network = network.to('xpu')
      network, optimizer = ipex.optimize(network, optimizer=optimizer, dtype=torch.bfloat16)
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        data = data.to("xpu")
        target = target.to("xpu")
        ....
        ...

```

**Comment by huhanwj** [comment]
> @huhanwj I could reproduce the original issue (related to max recursion) you had reported. Seems the issue was due to multiple invocation of ipex.optimize. Making the following change to the train function resolves the issue,
> 
> ```python
> def train(epoch):
>     global network
>     global optimizer
>     network.train()
>     if epoch == 1:
>         network = network.to('xpu')
>         network, optimizer = ipex.optimize(network, optimizer=optimizer, dtype=torch.bfloat16)
>     for batch_idx, (data, target) in enumerate(train_loader):
>         optimizer.zero_grad()
>         data = data.to("xpu")
>         target = target.to("xpu")
>         ....
>         ...
> ```

@vishnumadhu365 Thanks for your reply! However, make such changes will not update the network and then training for more epoches is just meaningless and will stuck at the same accuracy.

**Comment by jgong5** [comment]
> I could reproduce the original issue (related to max recursion) you had reported. Seems the issue was due to multiple invocation of ipex.optimize

This is also something we can further improve to avoid such an error when `ipex.optimize` is invoked multiple times. cc @zhuhaozhe 



> Thanks for your reply! However, make such changes will not update the network and then training for more epoches is just meaningless and will stuck at the same accuracy.

@huhanwj May I know why the proposed changes won't update the network?

**Comment by huhanwj** [comment]
> > I could reproduce the original issue (related to max recursion) you had reported. Seems the issue was due to multiple invocation of ipex.optimize
> 
> This is also something we can further improve to avoid such an error when `ipex.optimize` is invoked multiple times. cc @zhuhaozhe
> 
> > Thanks for your reply! However, make such changes will not update the network and then training for more epoches is just meaningless and will stuck at the same accuracy.
> 
> @huhanwj May I know why the proposed changes won't update the network?

It turns out that my network may have some issue with the trainning, change it to 
```python
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)
    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2,2,0))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2,2,0))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x)
```

training on CPU will give me decreasing loss and high accuracy from 91% (epoch 1) to 97% (epoch 5). However, parsing the data to XPU will always get a 11% accuracy no matter how many epoches you have trained.

Diff between using CPU and XPU:
on my 12900K I get
```
Train Epoch: 1 [0/60000 (0%)]	Loss: 2.358812
Train Epoch: 1 [1280/60000 (2%)]	Loss: 2.285138
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.306635
Train Epoch: 1 [3840/60000 (6%)]	Loss: 2.270880
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.236740
Train Epoch: 1 [6400/60000 (11%)]	Loss: 2.243347
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.183255
Train Epoch: 1 [8960/60000 (15%)]	Loss: 2.148512
Train Epoch: 1 [10240/60000 (17%)]	Loss: 2.104912
Train Epoch: 1 [11520/60000 (19%)]	Loss: 2.004407
Train Epoch: 1 [12800/60000 (21%)]	Loss: 1.862252
Train Epoch: 1 [14080/60000 (23%)]	Loss: 1.884371
Train Epoch: 1 [15360/60000 (26%)]	Loss: 1.797316
Train Epoch: 1 [16640/60000 (28%)]	Loss: 1.687911
Train Epoch: 1 [17920/60000 (30%)]	Loss: 1.508046
Train Epoch: 1 [19200/60000 (32%)]	Loss: 1.764279
Train Epoch: 1 [20480/60000 (34%)]	Loss: 1.470076
Train Epoch: 1 [21760/60000 (36%)]	Loss: 1.351451
Train Epoch: 1 [23040/60000 (38%)]	Loss: 1.290542
Train Epoch: 1 [24320/60000 (41%)]	Loss: 1.018779
Train Epoch: 1 [25600/60000 (43%)]	Loss: 1.044021
Train Epoch: 1 [26880/60000 (45%)]	Loss: 1.096556
Train Epoch: 1 [28160/60000 (47%)]	Loss: 1.229379
Train Epoch: 1 [29440/60000 (49%)]	Loss: 1.148369
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.842456
Train Epoch: 1 [32000/60000 (53%)]	Loss: 0.978211
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.858768
Train Epoch: 1 [34560/60000 (58%)]	Loss: 0.924494
Train Epoch: 1 [35840/60000 (60%)]	Loss: 1.018511
Train Epoch: 1 [37120/60000 (62%)]	Loss: 0.806291
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.878710
Train Epoch: 1 [39680/60000 (66%)]	Loss: 0.819970
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.802814
Train Epoch: 1 [42240/60000 (70%)]	Loss: 0.826412
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.854793
Train Epoch: 1 [44800/60000 (75%)]	Loss: 0.770171
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.720727
Train Epoch: 1 [47360/60000 (79%)]	Loss: 0.548549
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.756250
Train Epoch: 1 [49920/60000 (83%)]	Loss: 0.656857
Train Epoch: 1 [51200/60000 (85%)]	Loss: 0.700063
Train Epoch: 1 [52480/60000 (87%)]	Loss: 0.809167
Train Epoch: 1 [53760/60000 (90%)]	Loss: 0.786540
Train Epoch: 1 [55040/60000 (92%)]	Loss: 0.751619
Train Epoch: 1 [56320/60000 (94%)]	Loss: 0.539100
Train Epoch: 1 [57600/60000 (96%)]	Loss: 0.641221
Train Epoch: 1 [58880/60000 (98%)]	Loss: 0.660536

Test set: Avg. loss: 0.3244, Accuracy: 9099/10000 (91%)
```

on my A770 I get
```
Train Epoch: 1 [0/60000 (0%)]	Loss: 2.352571
Train Epoch: 1 [1280/60000 (2%)]	Loss: 2.342006
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.364790
Train Epoch: 1 [3840/60000 (6%)]	Loss: 2.359556
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.369886
Train Epoch: 1 [6400/60000 (11%)]	Loss: 2.335656
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.341964
Train Epoch: 1 [8960/60000 (15%)]	Loss: 2.300711
Train Epoch: 1 [10240/60000 (17%)]	Loss: 2.363933
Train Epoch: 1 [11520/60000 (19%)]	Loss: 2.326606
Train Epoch: 1 [12800/60000 (21%)]	Loss: 2.324704
Train Epoch: 1 [14080/60000 (23%)]	Loss: 2.379430
Train Epoch: 1 [15360/60000 (26%)]	Loss: 2.390502
Train Epoch: 1 [16640/60000 (28%)]	Loss: 2.286761
Train Epoch: 1 [17920/60000 (30%)]	Loss: 2.342486
Train Epoch: 1 [19200/60000 (32%)]	Loss: 2.383628
Train Epoch: 1 [20480/60000 (34%)]	Loss: 2.360693
Train Epoch: 1 [21760/60000 (36%)]	Loss: 2.285080
Train Epoch: 1 [23040/60000 (38%)]	Loss: 2.388155
Train Epoch: 1 [24320/60000 (41%)]	Loss: 2.339409
Train Epoch: 1 [25600/60000 (43%)]	Loss: 2.344594
Train Epoch: 1 [26880/60000 (45%)]	Loss: 2.338047
Train Epoch: 1 [28160/60000 (47%)]	Loss: 2.376128
Train Epoch: 1 [29440/60000 (49%)]	Loss: 2.396311
Train Epoch: 1 [30720/60000 (51%)]	Loss: 2.325555
Train Epoch: 1 [32000/60000 (53%)]	Loss: 2.337614
Train Epoch: 1 [33280/60000 (55%)]	Loss: 2.307302
Train Epoch: 1 [34560/60000 (58%)]	Loss: 2.374896
Train Epoch: 1 [35840/60000 (60%)]	Loss: 2.381089
Train Epoch: 1 [37120/60000 (62%)]	Loss: 2.361262
Train Epoch: 1 [38400/60000 (64%)]	Loss: 2.370956
Train Epoch: 1 [39680/60000 (66%)]	Loss: 2.383646
Train Epoch: 1 [40960/60000 (68%)]	Loss: 2.352216
Train Epoch: 1 [42240/60000 (70%)]	Loss: 2.333538
Train Epoch: 1 [43520/60000 (72%)]	Loss: 2.309168
Train Epoch: 1 [44800/60000 (75%)]	Loss: 2.334059
Train Epoch: 1 [46080/60000 (77%)]	Loss: 2.367992
Train Epoch: 1 [47360/60000 (79%)]	Loss: 2.358446
Train Epoch: 1 [48640/60000 (81%)]	Loss: 2.396939
Train Epoch: 1 [49920/60000 (83%)]	Loss: 2.333898
Train Epoch: 1 [51200/60000 (85%)]	Loss: 2.342280
Train Epoch: 1 [52480/60000 (87%)]	Loss: 2.340971
Train Epoch: 1 [53760/60000 (90%)]	Loss: 2.319399
Train Epoch: 1 [55040/60000 (92%)]	Loss: 2.350685
Train Epoch: 1 [56320/60000 (94%)]	Loss: 2.353016
Train Epoch: 1 [57600/60000 (96%)]	Loss: 2.377032
Train Epoch: 1 [58880/60000 (98%)]	Loss: 2.351476

Test set: Avg. loss: 2.3360, Accuracy: 1136/10000 (11%)
```

I don't know why how this happens. For the CPU training, I just comment all codes relating to "xpu".

**Comment by huhanwj** [comment]
Also one thing to mention is that though this simple network is modified from some code using CPU to train on MNIST, I constantly get plenty of lines stating the `[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware` error
also curious about where does this come from? Is it related to WSL terminal or what?

**Comment by vishnumadhu365** [comment]
@huhanwj Appreciate the extended tests and feedback. We will look into why the model doesn't converge while ran on the GPU.
Regarding the error "[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware", its due to the lack of FP64 support on the A770 and not specific to WSL. Did you get this error while running the same minst+CNN train experiment ?

**Comment by huhanwj** [comment]
> @huhanwj Appreciate the extended tests and feedback. We will look into why the model doesn't converge while ran on the GPU.
> Regarding the error `[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware`, its due to the lack of FP64 support on the A770 and not specific to WSL. Did you get this error while running the same MNIST+CNN train experiment ?

yes after training the epoch, the output of my terminal will burst a lot of such error, but pointing to the same kernel

both in FP16 and FP32

**Comment by vishnumadhu365** [comment]
@jingxu10 @jgong5  I could reproduce the reported observation 'training doesn't converge on gpu', needs further debugging

**Comment by gujinghui** [comment]
> > @huhanwj Appreciate the extended tests and feedback. We will look into why the model doesn't converge while ran on the GPU.
> > Regarding the error `[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware`, its due to the lack of FP64 support on the A770 and not specific to WSL. Did you get this error while running the same MNIST+CNN train experiment ?
> 
> yes after training the epoch, the output of my terminal will burst a lot of such error, but pointing to the same kernel
> 
> both in FP16 and FP32

As long as you saw this warning, it is impossible to get correct output then. This warning means, the corresponding operator/kernel has been poisoned/emptied with doing nothing on device, because this operator/kernel has FP64 computation.
In next solution, the warning will be replaced by an exception, which will block the host and stop model execution.

**Comment by huhanwj** [comment]
> > > @huhanwj Appreciate the extended tests and feedback. We will look into why the model doesn't converge while ran on the GPU.
> > > Regarding the error `[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware`, its due to the lack of FP64 support on the A770 and not specific to WSL. Did you get this error while running the same MNIST+CNN train experiment ?
> > 
> > yes after training the epoch, the output of my terminal will burst a lot of such error, but pointing to the same kernel
> > 
> > both in FP16 and FP32
> 
> As long as you saw this warning, it is impossible to get correct output then. This warning means, the corresponding operator/kernel has been poisoned/emptied with doing nothing on device, because this operator/kernel has FP64 computation.
> In next solution, the warning will be replaced by an exception, which will block the host and stop model execution.

Thanks for your clarification!

Then why will this happen even I tried to specify the data type as FP32/FP16 in ´ipex.optimize´? 

Or do I need to do further data type conversion before I parse the data to my xpu device?

This is weird to me as I only call some basic torch.nn functions.

**Comment by gujinghui** [comment]
> > > > @huhanwj Appreciate the extended tests and feedback. We will look into why the model doesn't converge while ran on the GPU.
> > > > Regarding the error `[CRITICAL ERROR] Kernel 'XXX' removed due to usage of FP64 instructions unsupported by the targeted hardware`, its due to the lack of FP64 support on the A770 and not specific to WSL. Did you get this error while running the same MNIST+CNN train experiment ?
> > > 
> > > 
> > > yes after training the epoch, the output of my terminal will burst a lot of such error, but pointing to the same kernel
> > > both in FP16 and FP32
> > 
> > 
> > As long as you saw this warning, it is impossible to get correct output then. This warning means, the corresponding operator/kernel has been poisoned/emptied with doing nothing on device, because this operator/kernel has FP64 computation.
> > In next solution, the warning will be replaced by an exception, which will block the host and stop model execution.
> 
> Thanks for your clarification!
> 
> Then why will this happen even I tried to specify the data type as FP32/FP16 in ´ipex.optimize´?
> 
> Or do I need to do further data type conversion before I parse the data to my xpu device?
> 
> This is weird to me as I only call some basic torch.nn functions.

Unfortunately, even no FP64 input is passed to operators, some operators will still trigger this warning, because:
1. According to PyTorch design, some kernel backbone will handle all data type in single kernel binary, including FP64.
2. In some case, Python frontend code in PyTorch will enforce to promote data type to double.

For these scenarios, we are going to work out solution soon.

**Comment by huhanwj** [comment]
issue fixed in new version 1.13.10+xpu
thank you all!!
